Dip/Dip/Definition.swift

```diff
@@ -50,20 +50,62 @@ func ==(lhs: DefinitionKey, rhs: DefinitionKey) -> Bool {
 public enum ComponentScope {
   /// Indicates that a new instance of the component will be created each time it's resolved.
   case Prototype
+  /// Indicates that instances will be reused during resolve but will be discurded when topmost `resolve` method returns.
+  case ObjectGraph
   /// Indicates that resolved component should be retained by container and always reused.
   case Singleton
 }
 
 ///Definition of type T describes how instances of this type should be created when they are resolved by container.
 public final class DefinitionOf<T>: Definition {
```

Comment:
Actually it worked and I kind of satisfied with result. You can see it in late commit for this PR.
Using additional generic type added some complexity for sure, but that also lets get rid of force casting factory (which was `Any`) in `resolve` method.
Another two very positive side effects of this refactoring is ability to register/remove individual dependencies. Without generic type of factory that would be impossible cause there will be no way to compare definitions (because factory is of type 'Any', so no way to use `==`).
Drawback is that now you have to separate creating definition and call to resolve dependencies like this:

```
    var serverDefinition = container.register(.ObjectGraph) { Server() as Server }
    serverDefinition.resolveDependencies(container) { container, server in
      server.client = container.resolve() as Client
    }
```

If you are ok with that change I will proceed with updating readme and playground.